server/src/main/java/org/elasticsearch/env/NodeEnvironment.java

```diff
@@ -470,6 +473,38 @@ private static boolean upgradeLegacyNodeFolders(Logger logger, Settings settings
         return true;
     }
 
+    /**
+     * Checks to see if we can upgrade to this version based on the existing index state. Upgrading
+     * from older versions can cause irreversible changes if allowed.
+     * @param logger
+     * @param nodePaths
+     * @throws IOException
+     */
+    static void checkForIndexCompatibility(Logger logger, NodePath... nodePaths) throws IOException {
+        final Path[] paths = Arrays.stream(nodePaths).map(np -> np.path).toArray(Path[]::new);
+        NodeMetadata metadata = PersistedClusterStateService.nodeMetadata(paths);
+
+        // We are upgrading the cluster, but we didn't find any previous metadata. Possibly corrupted state.
```

Comment:
Since we've confirmed by this point that we're really doing an upgrade and not just starting a new node, I think we should bail out with an exception rather than just assuming everything is ok.

Reply:
OK, sounds good.